### DISCUSSION TITLE
How to link identity with react native?

### DISCUSSION THREAD

**Discussion by f-bog** [supabase/supabase#25976] [Questions]
I'm currently integrating Apple authentication with Supabase Auth in a React Native app. I'm attempting to link anonymous users to an auth provider (Apple) using the linkIdentity method from Supabase Auth. This method should ideally trigger a redirect to the provider's authentication page. However, the documentation hasn't been very helpful in implementing this in a React Native environment.

[documentation](https://supabase.com/docs/guides/auth/auth-anonymous#link-an-oauth-identity)

Here’s the function I'm using to link anonymous auth with Apple:

```javascript
const linkAnonymousAuthWithApple = async () => {
      try {
        const { data, error } = await supabase.auth.linkIdentity({
          provider: "apple",
          options: {
            skipBrowserRedirect: true,
          },
        });

        if (error) throw error;
        console.log({ data });

        Linking.openURL(data.url); // from expo-linking
      } catch (err) {
        console.error(err);
      }
    };
```

Safari opens on the phone and displays a native sheet prompting for authentication details. However, after entering the required details, I encounter a problem. The sign-up process does not complete, and I'm left with a message indicating "Sign-up not completed.

Appreciate any help I can get.

**Comment by f-bog** [comment]
Not sure if i'm allowed to bump this? Really stuck on it. Would appreciate help

**Comment by pgregorynz** [comment]
I'm also looking for information on this?  Is it even possible?  Seems like it should be?

**Comment by chramos** [comment]
same issue

### edit:
I manage to solve the issue on React Native
for me the issue was that I wasn't using the nonce token in the signing part

```js
const appleAuthRequestResponse = await appleAuth.performRequest({
    requestedOperation: appleAuth.Operation.LOGIN,
    requestedScopes: [appleAuth.Scope.FULL_NAME, appleAuth.Scope.EMAIL],
});
await supabase.auth.signInWithIdToken({
    provider: 'apple',
    token: appleAuthRequestResponse.identityToken,
    nonce: appleAuthRequestResponse.nonce,
});
```

then, I was able to link the identity simply calling this

```js
await supabase.auth.linkIdentity({
    provider: 'apple',
    options: {skipBrowserRedirect: true},
});
```

**Comment by charlower** [comment]
Also stuck on this one, can't find a way to link an anonymous user to an apple account in react native.

**Comment by hayata-suenaga** [comment]
@chramos how did you call `supabase.auth.linkIdentity` from the same React Native app on which you did Apple authentication?

I was logged out from the app when I called `signInWithIdToken` and `linkIdentity` without any error thrown 🤔 

I'm using Google authentication.

```ts
    const userInfo = await GoogleSignin.signIn();
    await supabase.auth.signInWithIdToken({
      provider: "google",
      token: userInfo.idToken ?? "",
    });
    await supabase.auth.linkIdentity({
      provider: "google",
      options: { skipBrowserRedirect: true },
    });
```

**Comment by ducpt-bili** [comment]
+1 for that issue.
Hi @chramos , do your solution still work, i try it in my new project there is no performRequest or nonce i can pass in.
and calling signInWithIdToken will log out your anonymous user (not link).
Am i right or i missing something? 
Thanks.

**Comment by garrettwesley** [comment]
Anyone else figure out how to do this? If only linkIdentity would accept some authtoken.

**Comment by christophemenager** [comment]
It looks that there is no way to convert an anonymous user to an permanent user using native social sign in.

The only auth method that works to do this conversion is email.

I guess there is a manual workaround to store data in local storage > logout > sign In with social > upload new data and handle potential conflicts.

We would really need a way **to link identity with token**

**Comment by akodkod** [comment]
Guys, we need the ability to link identity by using an ID token.

We're building a React Native app where a user completes onboarding while logged in as an anonymous user (and we save information into our database based on this user). Then, when we authenticate them using OAuth and `signInWithIdToken`, we lose the anonymous user.

The ideal solution would be to have `linkIdentityWithIdToken` or something similar.

Thank you so much!

**Comment by alberthammerich** [comment]
Really need this feature too!

**Comment by timohuennebeck** [comment]
Same here. Can't seem to figure out how to do this without a link identity by ID token function.

Honestly, it's a bit unfortunate. We have a mobile app where users start off with an anonymous account, and we'd love to give them the option to authenticate via Apple after onboarding, so they don't have to deal with a sign-up screen right away. 

But for now, the only option we can offer them is authentication via OTP.

**Comment by bhkangw** [comment]
Is this being looked at by supabase??? Omfg at least add this massive limitation to the docs so people (like me) don't spend hours trying to figure out WHY it's not working

https://supabase.com/docs/guides/auth/auth-anonymous

**Comment by jerdmeyer** [comment]
+1 this is like the most important feature for auth right now because user's drop off early if they can't try the app without signing in first. But if we can't provide that option later on, then we can't allow them to save their existing content across devices easily.

**Reply by christophemenager** [reply]
I draw the same conclusions as you @jerdmeyer and I have the same needs. I have been waiting for 6 months for this issue to be solved but no one from supabase seems to be looking at this github issue

**Reply by jerdmeyer** [reply]
Agreed. I see this Github issue & response from what appears to be a Supabase dev - https://github.com/supabase/supabase-flutter/issues/1111#issuecomment-2603431844 but the issue has little to no traction and less visibility than this one (which has no Supabase dev commenting).

**Reply by timohuennebeck** [reply]
Yeah, I fully agree. This is such an important feature for my app, and it's really frustrating that it's not available yet.

I just sent Supabase an email about this and linked to this GitHub issue. Hopefully this helps get their attention.

**Reply by timohuennebeck** [reply]
Okay, so after having exchanged quite a few emails with the Supabase support team, I finally received a clear response from someone on the Supabase Auth team that this feature is indeed planned. While they couldn’t provide an exact ETA, it is confirmed to be on their roadmap 🙌

**Reply by dillontrembath** [reply]
Yes this feature is huge, very much needed.

**Comment by cwkeam** [comment]
+1, seems quite unbelievable this issue's been untracked for so long.

With:
```
if (anonymousSession) {
      // Link the anonymous account to the Google identity
      const { data, error } = await supabase.auth.linkIdentity({
          provider: "google",
      });
```
For me, I get `{"provider": "google", "url": "https:https://accounts.google.com/o/oauth2/v2/auth?client_id..."}`, which makes me think this is probably a feature that was only meant for web-based applications.

**Comment by ksawery29** [comment]
After a solid while of reverse engineering supabase, I've figured it out. My need for it was to link OAuth to an anonymous user so i'll focus on that, however, it should be pretty straight forward for already authenticated users too.

First, you will do your OAuth stuff as you do normally on the client, for example, here's in expo + apple authentication:
```js
const credential = await AppleAuthentication.signInAsync({
  requestedScopes: [
    AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
    AppleAuthentication.AppleAuthenticationScope.EMAIL,
  ],
});
```

What matters here is that the user finishes the authentication flow and give us their identity token. We do not need anything else right now. In expo, this can be done as so:

```js
const identityToken = credential.identityToken
```

Now the hard part: linking **anonymous users**.  For this, you will need to have a custom server. It can either be aws, railway, supabase or anything else - it doesn't really matter as long as the end user doesn't see the code.

On the server, you will need to gather two things:
- User ID
- Email

**Do not send and receive the email from the body - as anyone could put any email and cause massive security issues**. Instead, looking at identity token that apple sends us (possibly with Google it's the same - though I haven't tested it) it is a JWT, and after decoding it we securely can get the email. So, great, now we have a safe way to pass user data.
With that in mind, on the client side make a fetch request to your server **only passing the user id and identity token**.

Now let's go back to the server - any action below that will also be executed there.
First, you should verify the identity token that the user sends. With nodejs, this can be done using the `jwks-rsa` & `jsonwebtoken` libraries.  The implementation is pretty straight-forward:

Start by getting the public key:
```js
import jwksClient from "jwks-rsa";
async function getPublicKey(kid: string, platform: "ios" | "android") {
  const jwks = jwksClient({
    jwksUri: platform === "ios" ? "https://appleid.apple.com/auth/keys" : "https://www.googleapis.com/oauth2/v3/certs",
  });
  const key = await jwks.getSigningKey(kid);
  return key.getPublicKey();
}
```

And then decode it using `jwt`:
```js
import { decode, verify } from "jsonwebtoken";

const publicKey = await getPublicKey(decoded.kid, "ios" /* or android */);
const verified = verify(identityToken, publicKey, { algorithms: ["RS256"] });
if (!verified) {
  throw new Error("Invalid identity token");
}
```

Note that you will need to decode the jwt token before. This can be done also using the `jsonwebtoken` library by simply using `jwt.decode`. This will also be the place where you will get the email. The email is usually stored in `payload.email`.

Now we have 100% certainty that the token was actually signed by Apple or Google. Next step is to link the OAuth email to the user.
Using the supabase REST api, you can do the following request to your hosted instance:

```
PUT https://<YOUR SUPABASE URL>/auth/v1/admin/users/<THE USER ID YOU WANT TO MODIFY>
Headers:
  apiKey: <YOUR SUPABASE SERVICE ROLE KEY>
  Authorization: <YOUR SUPABASE SERVICE ROLE KEY>
Body:
  {"email": <EMAIL FROM THE IDENTITY TOKEN>, "email_confirm": true}
```

or, if you are using the supabase sdk with service role key:
```js
const { error } = await supabaseClient.auth.admin.updateUserById(userId, {
  email,
  email_confirm: true,
});
```

This will set the same email as the user's Apple Account. It is required for later.

Notice how in the request we set `email_confirm` to true. This is for a reason. Otherwise, the user would get an email confirmation request which is not the purpose of OAuth sign ups/ins. This is also the reason why it's done on the server - although you can also add email to the user using client side code, you cannot automatically verify their email there.

Our authenticated user now has email added which is the same which apple gave us - meaning we are halfway there. Great.
This is also the end of your server side logic. Rest will be done on the client.

Referring to supabase docs, we see the following:
> Supabase Auth automatically links identities with the same email address to a single user. (source: https://supabase.com/docs/guides/auth/auth-identity-linking?queryGroups=language&language=js)

Which means, now that the anonymous user has email connected, we can sign in with the token using the supabase SDK - and supabase will automatically detect our anonymous user and link the oauth connection.

Signing in with the oauth token is straight-forward. In expo, just do:
```js
const { data, error } = await supabase.auth.signInWithIdToken({
  provider: "apple",
  token: credential.identityToken,
});
```

And that's it. Supabase *should* link the identity. It is a mess - and had me questioning what I pay for, but whatever. Note that you most likely want to secure your server with some additional verification, otherwise an attacker could mess your users up.

**Reply by f-bog** [reply]
Thanks for this. Seems pretty solid. Just bewildered why this isn't something offered out of the box.

**Comment by cwkeam** [comment]
> After a solid while of reverse engineering supabase, I've figured it out. My need for it was to link OAuth to an anonymous user so i'll focus on that, however, it should be pretty straight forward for already authenticated users too.
> 
> First, you will do your OAuth stuff as you do normally on the client, for example, here's in expo + apple authentication:
> 
> ```js
> const credential = await AppleAuthentication.signInAsync({
>   requestedScopes: [
>     AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
>     AppleAuthentication.AppleAuthenticationScope.EMAIL,
>   ],
> });
> ```
> 
> What matters here is that the user finishes the authentication flow and give us their identity token. We do not need anything else right now. In expo, this can be done as so:
> 
> ```js
> const identityToken = credential.identityToken
> ```
> 
> Now the hard part: linking **anonymous users**. For this, you will need to have a custom server. It can either be aws, railway, supabase or anything else - it doesn't really matter as long as the end user doesn't see the code.
> 
> On the server, you will need to gather two things:
> 
> * User ID
> * Email
> 
> **Do not send and receive the email from the body - as anyone could put any email and cause massive security issues**. Instead, looking at identity token that apple sends us (possibly with Google it's the same - though I haven't tested it) it is a JWT, and after decoding it we securely can get the email. So, great, now we have a safe way to pass user data. With that in mind, on the client side make a fetch request to your server **only passing the user id and identity token**.
> 
> Now let's go back to the server - any action below that will also be executed there. Using the supabase REST api, you can do the following request to your hosted instance:
> 
> ```
> PUT https://<YOUR SUPABASE URL>/auth/v1/admin/users/<THE USER ID YOU WANT TO MODIFY>
> Headers:
>   apiKey: <YOUR SUPABASE SERVICE ROLE KEY>
>   Authorization: <YOUR SUPABASE SERVICE ROLE KEY>
> Body:
>   {"email": <EMAIL FROM THE IDENTITY TOKEN>, "email_confirm": true}
> ```
> 
> or, if you are using the supabase sdk with service role key:
> 
> ```js
> const { error } = await supabaseClient.auth.admin.updateUserById(userId, {
>   email,
>   email_confirm: true,
> });
> ```
> 
> This will set the same email as the user's Apple Account.
> 
> Notice how in the request we set `email_confirm` to true. This is for a reason. Otherwise, the user would get an email confirmation request which is not the purpose of OAuth sign ups/ins. This is also the reason why it's done on the server - although you can also add email to the user using client side code, you cannot automatically verify their email there.
> 
> If you don't know how to get the email from the identity token, here's how to do it in bun:
> 
> * Install the `jsonwebtoken` package using bun add ...
> * And on your server run `jwt.decode` of course importing jwt from jsonwebtoken.
> 
> On Apple OAuth, the email is sent as: `payload.email`.
> 
> Our authenticated user now has email added which is the same which apple gave us - meaning we are halfway there. Great. This is also the end of your server side logic. Rest will be done on the client.
> 
> Referring to supabase docs, we see the following:
> 
> > Supabase Auth automatically links identities with the same email address to a single user. (source: https://supabase.com/docs/guides/auth/auth-identity-linking?queryGroups=language&language=js)
> 
> Which means, now that the anonymous user has email connected, we can sign in with the token using the supabase SDK - and supabase will automatically detect our anonymous user and link the oauth connection.
> 
> Signing in with the oauth token is straight-forward. In expo, just do:
> 
> ```js
> const { data, error } = await supabase.auth.signInWithIdToken({
>   provider: "apple",
>   token: credential.identityToken,
> });
> ```
> 
> And that's it. Supabase _should_ link the identity. It is a mess - and had me questioning what I pay for, but whatever. Note that you most likely want to secure your server with some additional verification, otherwise an attacker could mess your users up.


Wondering about the following:

Identity token is signed by google/apple right? So we don’t have the JWT secret to verify whether it came from google or apple (or others). It seems with this approach, I could sign an arbitrary jwt containing a fake email and it’d still put me through.

We can make a case whether this matters - since ultimately all users sign in via oauth, you couldn’t log into my personal google account-linked supabase account with this method anyway I think. That is to say that the proposed backend endpoint with JWT as input seems equally as secure as accepting userid and email. The essence of security seems to be in the front-end’s ability to generate the identity token via signing in via google/apple, and it seems quite secure!

Apple private relay I’m hoping would work well also, since in the end Apple just maps your apple email to a private email, and supabase just recognizes the user with the private email right?

Thanks for this idea @ksawery29, I think this would work. Although I do hope supabase just has a linkAccountWithIdentityToken function. Would be happy to try a PR if supabase team’s open to it

**Reply by ksawery29** [reply]
Good catch! Looks like you actually can verify the token. Although not tested, it seems like both Apple and Google have their public keys listed here: [Google](https://www.googleapis.com/oauth2/v3/certs), [Apple](https://appleid.apple.com/auth/keys).

So you can find the correct key by the `kid` and have the token verified. 

But you are right, a built in solution would be a lot better - and potentially safer. I wrote it because I needed something asap.

Once I actually try it I'll modify the comment I've made with this fix.

**Comment by Mykyta-Chernenko** [comment]
I think I figured it out! this is how it can be achieved much simpler. 

```
const newUser = await supabase.auth.signInAnonymously();
```
....
then after some time when I do a social login with google
```
const userInfo = await GoogleSignin.signIn();
if (userInfo.data.user.email){ 
  const { data: updateEmailData, error: updateEmailError } = await supabase.auth.updateUser(
      {
        email: userInfo.data.user.email,
      },
    );
}
 
const { data, error } = await supabase.auth.signInWithIdToken({
    provider: 'google',
    token: userInfo.data.idToken!,
  });
```
or with apple
```
const credential = await AppleAuthentication.signInAsync({
  requestedScopes: [
    AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
    AppleAuthentication.AppleAuthenticationScope.EMAIL,
  ],
});
if (credential.email) {
  console.log(credential.email);
  const { data: updateEmailData, error: updateEmailError } = await supabase.auth.updateUser(
    {
      email: credential.email,
    },
  );
}
const {
  error,
  data: { user },
} = await supabase.auth.signInWithIdToken({
  provider: 'apple',
  token: credential.identityToken,
});  
```
so the idea is for a social login
1) do anon sign in
2) do any sign in you want, social or email. 
3) get email from the social data, and associate it with the anon user
4) now sign in using the new social user just like before
5) because the email is the same, you will login to the "anon" account, which will be converted to a normal account

**Reply by bpeck81** [reply]
This doesnt work because updating the email doesnt update it in the supabase auth table and instead updates a field called email_change until the user autheticates the email, presumably through a verification link sent to their email.

**Comment by bpeck81** [comment]
Here's the copiable supabase edge function code to do what @ksawery29 suggests above.

In essence, you just:
1. deploy an edge function (I called it social-auth-link) with the code below
2. get the token from your provider
3. call the edge function with the token
4. sign in with the token as usual

For those joining, pretty much you need to set the anon user's email to the social provider's email before you sign in with supabase. Supabase has "automatic linking", which links accounts via email and this takes advantage of that.


Client Side:
```
 const credential = await AppleAuthentication.signInAsync({
                                    requestedScopes: [
                                        AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
                                        AppleAuthentication.AppleAuthenticationScope.EMAIL,
                                    ],
                                })
if (credential.identityToken) {
        const userId = (await supabase.auth.getSession()).session.user.id
        const { data, error } = await supabase.functions.invoke("social-auth-link", {
        body: { userId, identityToken, provider: "apple" }
        });
const {error,data: { user },} = await supabase.auth.signInWithIdToken({
        provider: 'apple',
       token: credential.identityToken,        
})

```



Edge function (social-auth-link):
[See here for the super secure version
](https://chatgpt.com/share/68092e3b-0c08-8008-8aaf-bcae22e7b6d7)
```
// src/functions/link-oauth/index.ts
import { serve } from "https://deno.land/std@0.178.0/http/server.ts";
import { createClient } from "https://esm.sh/@supabase/supabase-js@2";
import { jwtVerify, createRemoteJWKSet } from "https://esm.sh/jose@4";

const SUPABASE_URL = Deno.env.get("SUPABASE_URL")!;
const SERVICE_ROLE_KEY = Deno.env.get("SUPABASE_SERVICE_ROLE_KEY")!;

const supabaseAdmin = createClient(SUPABASE_URL, SERVICE_ROLE_KEY, {
  auth: { persistSession: false }
});

const CORS_HEADERS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "POST, OPTIONS",
  "Access-Control-Allow-Headers": "Content-Type, Authorization",
};

serve(async (req) => {
  // Always handle CORS preflight
  if (req.method === "OPTIONS") {
    return new Response(null, { headers: CORS_HEADERS });
  }
  if (req.method !== "POST") {
    return new Response("Method Not Allowed", {
      status: 405,
      headers: CORS_HEADERS,
    });
  }

  let body: { userId?: string; identityToken?: string; provider?: string };
  try {
    body = await req.json();
  } catch {
    return new Response("Invalid JSON", {
      status: 400,
      headers: CORS_HEADERS,
    });
  }

  const { userId, identityToken, provider } = body;
  if (!userId || !identityToken || !provider) {
    return new Response("Missing userId, identityToken or provider", {
      status: 400,
      headers: CORS_HEADERS,
    });
  }

  // pick the right JWKS endpoint
  let jwksUri: string;
  if (provider === "apple") {
    jwksUri = "https://appleid.apple.com/auth/keys";
  } else if (provider === "google") {
    jwksUri = "https://www.googleapis.com/oauth2/v3/certs";
  } else {
    return new Response("Unsupported provider", {
      status: 400,
      headers: CORS_HEADERS,
    });
  }

  try {
    const JWKS = createRemoteJWKSet(new URL(jwksUri));
    const { payload } = await jwtVerify(
      identityToken,
      JWKS,
      { algorithms: ["RS256"] }
    );

    const email = (payload as any).email;
    if (!email) {
      return new Response("Token did not contain email", {
        status: 400,
        headers: CORS_HEADERS,
      });
    }

    const { error } =
      await supabaseAdmin.auth.admin.updateUserById(userId, {
        email,
        email_confirm: true,
      });

    if (error) {
      console.error("Supabase updateUserById error:", error);
      return new Response(
        JSON.stringify({ error: error.message }),
        {
          status: 500,
          headers: { ...CORS_HEADERS, "Content-Type": "application/json" },
        }
      );
    }

    return new Response(
      JSON.stringify({ success: true }),
      {
        status: 200,
        headers: {
          ...CORS_HEADERS,
          "Content-Type": "application/json",
        },
      }
    );
  } catch (err) {
    console.error("Link OAuth error:", err);
    return new Response(
      JSON.stringify({ error: (err as Error).message }),
      {
        status: 500,
        headers: {
          ...CORS_HEADERS,
          "Content-Type": "application/json",
        },
      }
    );
  }
});

```

**Comment by dillontrembath** [comment]
Any update on the timeline for this feature?

**Comment by gitStanda** [comment]
I would also appreciate this.

**Comment by aneym** [comment]
bumping

**Comment by loryhuz** [comment]
I was migrating my app from Firebase auth and just got into that issue as well.

**Comment by tan-loc-nguyen** [comment]
run into this problem as well 😢 bumpinggggg

**Comment by MaxTheTurtle0** [comment]
Also running into this issue. Imo there should at least be something in the docs linking to this issue or outlining the limitation.

**Comment by ksawery29** [comment]
They’re too busy adding supabase to vibe coding platforms 😂

**Comment by bgdom** [comment]
They added the linking with idToken at least on the version 2.78.0 when i'm writing this. User `linkIdentity()` and pass the provider and token and maybe other options
